### DISCUSSION TITLE
Operons in JBrowse2

### DISCUSSION THREAD

**Discussion by sgiorgetti** [GMOD/jbrowse-components#4055] [Q&A]
Hello guys,

I have a question about operon's representation in JBrowse2.

If I load a track from GFF3 file in which operons are "standalone" features - say something like
```
contig	C1	operon	6366331	6369596	.	+	.	genes=gene1,gene2;Name=operon1;ID=operon1
contig	C1	operon	6441644	6446465	.	+	.	genes=gene3,gene4;Name=operon2;ID=operon2
```
All looks good and I can see the features displayed in my (linear) view.

If instead I describe the operon like 
```
contig	C1	operon	6366331	6369596	.	+	.	genes=gene1,gene2;Name=operon1;ID=operon1
contig	C1	gene	6366331	6366431	.	+	.	ID=gene1;Name=mygene1;Parent=operon1
contig	C1	gene	6368096	6369096	.	+	.	ID=gene2;Name=mygene2;Parent=operon1
```
it does not get rendered ... nor its gene subfeatures.

Am I doing anything wrong? May you kindly advise?

**Comment by cmdcolin** [comment]
unfortunately I think JBrowse 2 (and JBrowse 1 for that matter) may not handle operons super-well, however, the fact that you are not seeing it render at all is odd.

I tried copying your second example, and I was able to see it render though

![image](https://github.com/GMOD/jbrowse-components/assets/6511937/8e49f663-1568-47bd-8fdc-dce1b83c9cda)

i can't really be sure why you aren't seeing anything at all...the most common reason for things not rendering at all is that the 'refnames' don't match (first column of gff with names in fasta file) but not sure that applies in your case

**Reply by cmdcolin** [reply]
note: i adjusted coordinates and refnames just to load into our sample volvox example but that's all

**Reply by sgiorgetti** [reply]
Hello @cmdcolin  - thanks. I might have more info for you.

**Good news**: I can replicate "your" case and could see the operon.
**Bad news**: I cannot see "my" operon. Sort of.

![Screenshot 2023-11-14 at 17 39 21](https://github.com/GMOD/jbrowse-components/assets/78080390/13cfb377-5f38-46ba-87a9-51e53a0ff6b3)

Reason is I borrowed and adapted the 1st snippet above from a real GFF3 files and, it turned out, I adapted too much.
The nasty thing I have is an operon with 2 genes, which also overlap (coords are real ones this time; just renamed features)
```
contig	C1	operon	664	1426	.	-	.	ID=operon1;name=test_operon;genes=gene1,gene2
contig	C1	gene	664	855	.	-	0	ID=gene1;Parent=operon1
contig	C1	gene	842	1426	.	-	0	ID=gene2;Parent=operon1
contig	C1	mRNA	664	1426	.	-	0	ID=transcript1;Parent=gene1,gene2
contig	C1	exon	664	1426	.	-	0	ID=exon1;Parent= transcript1
contig	C1	CDS	664	855	.	-	0	ID=CDS1;Parent=transcript1;Derives_from=gene1
contig	C1	CDS	842	1426	.	-	0	ID=CDS2;Parent=transcript1;Derives_from=gene2
```

With this snippet, I cannot see the operon. If I select it, I can see the azure border of it, but no glyphs are rendered.
Also, the rectangle starts at the expected coordinate 664, but ends in the 3,000+ ballpark.

If I remove the CDS features, I can see some glyphs rendered outside the operon/rectangle, which is still overstretched (see image above).

If I remove the exon, it gets "better": the operon/rectangle looks good (see below)
![Screenshot 2023-11-14 at 17 58 23](https://github.com/GMOD/jbrowse-components/assets/78080390/c73adf5c-0ada-494b-9fc9-9cf1953fd559)

Finally, removing the transcript resolves the rendering issue, and I could see the operon along with its genes.

_note:_ Artificially hacking `gene2` to remove the overlap in the first place, does not fix the rendering issue.

Hope this helps.

**Reply by sgiorgetti** [reply]
If confirmed, I suppose this is a bug, which would end the discussion.

**Reply by cmdcolin** [reply]
what would you think the expected behavior might be for JBrowse to do to handle this properly? JBrowse could potentially 'ignore' the operon level of the feature during rendering maybe, at least as a starting step maybe

**Reply by cmdcolin** [reply]
it could even be that our data adapter strips the operon level of the feature. as nice as the "generalized" nature of the GFF3 child->parent hierarchy is, there often are assumptions for how those hierarchies need to be rendered by e.g. JBrowse that needs lots of hard coded stuff for e.g. how to draw a transcript, how to draw a exon subfeature of a transcript, etc. (at least how we have it currently! it would be great if the rendering could be more generalized too!)

**Reply by sgiorgetti** [reply]
Ignore the operon level could be a convenient and temporary solution from system/dev perspective, but researchers would benefit (i.e. need) the operon to be rendered somehow.
From system's point of view, the operon is a group of genes and stays one level up the hierarchy, as you mentioned.

During a workshop for gathering requirements for Apollo3, users fancied genes in an operon could be coloured the same. Not sure if this helps or not (guess not).
Shall have a chat with my JBrowse2/Apollo3 developer - and some users too - to see if we get a "decent" idea to share with you. Happy to help, as sorting this would be great to us too :-)

**Reply by sgiorgetti** [reply]
@cmdcolin - apologies for my belated follow up.

Had some discussions around, which I can summarise as follows, albeit not bringing brilliant ideas or elements to the discussion:
- An operon is a group of genes, close to each other, as far as the current/official understanding goes.
- These genes should be co-regulated, but this part is not captured by the GFF3 specs, yet - EOY 2023.
- Colour coding is nice from user perspective, **but** it would probably require some further _ad-hoc hardcoding_ or _spurious attributes_ in the GFF3 9th field. I would therefore suggest to keep the current rendering (rectangle enclosing the genes).
- Could not wrap my head around "rules" for paving the way to a "more generalised rendering of the features".
- This operon thing is something under consideration in the Apollo context. Not sure if there are synergies worth exploring.
